tests/v2/validation/provisioning/rke2/etcd_bkp_restore_test.go

```diff
@@ -0,0 +1,383 @@
+
+	logrus.Infof("fetching deployment list to validate restore.............")
+	deploymentList, err := steveclient.SteveType(workloads.DeploymentSteveType).List(&types.ListOpts{
+		Filters: map[string]interface{}{
+			"fieldSelector": "metadata.namespace=" + r.ns,
+		}})
+	require.NoError(r.T(), err)
+	require.Equal(r.T(), 1, len(deploymentList.Data))
+	require.Equal(r.T(), w1Name, deploymentList.Data[0].ObjectMeta.Name)
+	logrus.Infof(" deployment list validated successfully.............")
+
+	logrus.Infof("fetching ingresses list to validate restore.............")
+	ingressResp, err := steveclient.SteveType(ingresses.IngressSteveType).ByID(r.ns + "/" + ingress1.Name)
+	require.NoError(r.T(), err)
+	require.NotNil(r.T(), ingressResp)
+	logrus.Infof("ingress validated successfully.............")
+
+	cluster, _, err = getProvisioningClusterByName(client, clusterName, r.ns)
+	require.NoError(r.T(), err)
+	require.Equal(r.T(), initialK8sVersion, cluster.Spec.KubernetesVersion)
+}
+
+func (r *RKE2EtcdSnapshotRestoreTestSuite) watchAndWaitForCluster(kubeProvisioningClient *kubeProvisioning.Client, clusterName string) {
+	logrus.Infof("waiting for cluster to be up.............")
+	result, err := kubeProvisioningClient.Clusters(r.ns).Watch(context.TODO(), metav1.ListOptions{
+		FieldSelector:  "metadata.name=" + clusterName,
+		TimeoutSeconds: &defaults.WatchTimeoutSeconds,
+	})
+	require.NoError(r.T(), err)
+
+	checkFunc := clusters.IsProvisioningClusterReady
+	err = wait.WatchWait(result, checkFunc)
+	require.NoError(r.T(), err)
+}
+
+func (r *RKE2EtcdSnapshotRestoreTestSuite) watchAndWaitForDeployment(client *rancher.Client, clusterID string, deploymentName string) {
+
+	logrus.Infof("waiting for deployment to be created.............")
+
+	err := kwait.Poll(5*time.Second, 5*time.Minute, func() (done bool, err error) {
+		steveclient, err := client.Steve.ProxyDownstream(clusterID)
+		if err != nil {
+			return false, nil
+		}
+		deploymentResp, err := steveclient.SteveType(workloads.DeploymentSteveType).ByID(r.ns + "/" + deploymentName)
+		if err != nil {
+			return false, nil
+		}
+		deployment := &appv1.Deployment{}
+		err = provisioningV1.ConvertToK8sType(deploymentResp.JSONResp, deployment)
+		if err != nil {
+			return false, nil
+		}
+		if *deployment.Spec.Replicas == deployment.Status.AvailableReplicas {
+			return true, nil
+		}
+		return false, nil
+
+	})
+
+	require.NoError(r.T(), err)
+}
+
+func (r *RKE2EtcdSnapshotRestoreTestSuite) watchAndWaitForPods(client *rancher.Client, clusterID string) {
+	logrus.Infof("waiting for all Pods to be up.............")
+	err := kwait.Poll(5*time.Second, 5*time.Minute, func() (done bool, err error) {
+		steveClient, err := client.Steve.ProxyDownstream(clusterID)
+		if err != nil {
+			return false, nil
+		}
+		pods, err := steveClient.SteveType(pods.PodResourceSteveType).List(&types.ListOpts{})
+		if err != nil {
+			return false, nil
+		}
+		if len(pods.Data) == 0 {
+			return false, nil
+		}
```

Comment:
This is a nitpick, I definitely get your point for this length check, in Go, `range` won't iterate below for loop if this slice is empty or nil.

Reply:
Yeah right, I removed it since it's not needed.